types/basetypes/dynamic_value.go

```diff
@@ -0,0 +1,151 @@
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-go/tftypes"
+)
+
+var (
+	_ DynamicValuable = DynamicValue{}
+)
+
+// DynamicValuable extends attr.Value for dynamic value types. Implement this interface
+// to create a custom Dynamic value type.
+type DynamicValuable interface {
+	attr.Value
+
+	// ToDynamicValue should convert the value type to a DynamicValue.
+	ToDynamicValue(context.Context) (DynamicValue, diag.Diagnostics)
+}
+
+// DynamicValuableWithSemanticEquals extends DynamicValuable with semantic equality logic.
+type DynamicValuableWithSemanticEquals interface {
+	DynamicValuable
+
+	// DynamicSemanticEquals should return true if the given value is
+	// semantically equal to the current value. This logic is used to prevent
+	// Terraform data consistency errors and resource drift where a value change
+	// may have inconsequential differences.
+	//
+	// Only known values are compared with this method as changing a value's
+	// state implicitly represents a different value.
+	DynamicSemanticEquals(context.Context, DynamicValuable) (bool, diag.Diagnostics)
+}
+
+// NewDynamicValue creates a Dynamic with a known value. Access the value via the Dynamic
+// type UnderlyingValue method. The concrete value type returned to Terraform from this value
+// will be determined by the provided `(attr.Value).ToTerraformValue` function.
+func NewDynamicValue(value attr.Value) DynamicValue {
+	return DynamicValue{
+		value: value,
+		state: attr.ValueStateKnown,
+	}
+}
+
+// NewDynamicNull creates a Dynamic with a null value. The concrete value type returned to Terraform
+// from this value will be tftypes.DynamicPseudoType.
+func NewDynamicNull() DynamicValue {
+	return DynamicValue{
+		state: attr.ValueStateNull,
+	}
+}
+
+// NewDynamicUnknown creates a Dynamic with an unknown value. The concrete value type returned to Terraform
+// from this value will be tftypes.DynamicPseudoType.
+func NewDynamicUnknown() DynamicValue {
+	return DynamicValue{
+		state: attr.ValueStateUnknown,
+	}
+}
+
+// DynamicValue represents a dynamic value. Static types are always
+// preferable over dynamic types in Terraform as practitioners will receive less
+// helpful configuration assistance from validation error diagnostics and editor
+// integrations.
+type DynamicValue struct {
+	// value contains the known value, if not null or unknown.
+	value attr.Value
+
+	// state represents whether the value is null, unknown, or known. The
+	// zero-value is null.
+	state attr.ValueState
+}
+
+// Type returns DynamicType.
+func (v DynamicValue) Type(ctx context.Context) attr.Type {
+	return DynamicType{}
+}
+
+// ToTerraformValue returns the equivalent tftypes.Value for the DynamicValue.
+func (v DynamicValue) ToTerraformValue(ctx context.Context) (tftypes.Value, error) {
+	switch v.state {
+	case attr.ValueStateKnown:
+		return v.value.ToTerraformValue(ctx)
```

Comment:
Should I double check this isn't `nil`? Return a more formal error? Return a valid dynamic nil?

Reply:
Oof, yes, I believe we should return an error in that situation. We should try to avoid converting value states "automatically" since it may be an implementation error that leads to confusing errors later like Terraform's inconsistent result after apply error. Ideally, we can try our best to prevent this from being a possible situation in other methods/constructor functions.

Reply:
> Can this ever be nil if v.state is ValueStateKnown

After we fix up the `NewDynamicValue` function, it will be impossible without explicitly doing something like:
```go
imLookingForABadTime := types.Dynamic{
  value: nil,
  state: attr.ValueStateKnown,
}
```

With the changes in 6bb2360 we will make sure we never panic even if they did the above ⚡